Create work experience editor

jaspax opened the pull request:
**Release Type:** *Non-Breaking feature* 

Fixes https://x-team-internal.atlassian.net/browse/XP-3736

## Description

Implement a `WorkExperienceEntryEditor` component, which allows for condensed editing of work experience entries. This component is intended for integration in the XP Admin dashboard, and so is briefer and has less extensive validation than the applicant-facing work experience page. However, many of the same components are reused.

## Checklist

- [x] set yourself as an assignee
- [x] set appropriate labels for a PR (`In Review` or `In Progress` depending on its status)
- [x] move respective JIRA issue to the `PEER REVIEW` column
- [x] Flow typechecks passed (`npm run flow`)
- [x] Snapshots tests passed (`npm run jest`) (**NOTE:** This is temporarily disabled. Skip it.)
- [x] if added a new UI component that is meant to be used directly by other apps, remember to export that from the appropriate bundle file: `src/index.js` and/or `src/registration.js`
- [x] component's documentation (`.stories.js` file) is changed or added accordingly to reflect any new or updated use cases and/or debug cases
- [x] if any snapshots have been changed, verify that component still works and looks as expected and update the changed snapshot (**NOTE:** This is temporarily disabled. Skip it.)
- [x] **manually tested the app** by running it in several different browsers (Firefox, Chrome, Opera, Safari, MS IE/Edge, etc.) and checked nothing is broken and operates as expected!

## Steps to Test or Reproduce

Open Storybook and navigate to UI Components > Work Experience > Editor. This will show you the edit view on the work experience form. Validate that you can edit and work with all controls as expected.

In UI Components > Work Experience > "basic usage" you can test the combined Presenter/Editor function. The storybook has a stub implementation that allows you to edit an experience entry inline and then save it, and you will see new entries added to the list. You can also click the "Add New Experience" button at the bottom of the list to add an entirely new entry.

You may also look at the console and observe the object that is passed to the `onSave` method after you click "Save". The storybook implementation logs this object to the console so that you can validate that all modified fields are passed to the `onSave` method.

Note that in the "No experience provided" case, you can't actually click the "Add New Experience" button and have your newly-created experience be added. This is a shortcoming of the Storybook stub, not the control itself.

## Impacted Areas in Application

XP Admin interface.

## Screenshots

![image](https://user-images.githubusercontent.com/5854132/81071495-79cb9580-8eed-11ea-9457-fd0b8ea78ff6.png)

jaspax commented:
Note: there's a known issue here where newlines are being lost in the textarea. I'll be looking at this tomorrow.

jaspax commented:
Also please note that I included the checkins from the branch `XP-3735-create-work-experience-presenter`, because there was some shared work. I will hopefully merge that branch first, so some of the file diffs here will go away.

bernardodiasc commented on src/components/ui/WorkExperience/WorkExperienceEntryEditor.js:
```diff
@@ -0,0 +1,112 @@
+      </div>
+      <div className={cx.highlights}>
+        <p>Highlights</p>
+        <TextareaEditor text={highlights || ''} placeholder='Highlights' charLimit={1000}
```
It should be safe to use `<RichTextEditor>` here, as long it saves HTML data :)

jaspax commented:
@dennermiranda I fixed that, a simple mistake on the stories page.

dennermiranda commented:
> @dennermiranda I fixed that, a simple mistake on the stories page.

Thanks, @jaspax, I'm going to take another look

nicksp commented:
Hey @jaspax Please resolve merge conflicts when you have a moment.

kararade commented:
> Hey @jaspax Please resolve merge conflicts when you have a moment.

hey guys, I've fixed the conflicts 😉 this is ready to review again 👍

kararade commented:
I have verified the code and it looks good to me, I have also tested it by following the instructions on my local machine and everything works as expected 👍

jaspax commented:
> 1. Labels use is different

I removed the "date" label and changed "Skills" to "Tech stack"

> Also, notice, that Start date and final date fields should occupy the whole width and not be fixed limited.

The date fields already occupy the whole width so I'm not sure what you want here.

> 2. During the last discussions, we seemed to agree that components won't have a red focus highlight as it's misleading. Bernardo fixed that for Select components in his recent PR.

I'm reusing the `SkillsSelector` and `Timeframe` controls specified elsewhere, so once we merge into a branch which has Bernardo's updates you should see them here as well.

> 3. From designs, "Highlight" field should be a simple text area and doesn't contain any toolbars as we can see in current changes. To achieve that, I recommend we use TextareaEditor component unless it possesses any limitations.

The design document _does_ show bulleted lists and other basic formatting, implying rich text capability. The control that we have which supports that is `RichTextEdit`. I'll check with Paul to make sure this is okay.

jaspax commented:
@nicksp Updated after talking with Paul. The `Highlights` field was converted to a plain text edit. In order to make this work properly on this layout I had to remove the `max-width` property from the `TextareaEdit`. I checked the other places where we are using this control and they all have the field confined to a column which sets the width appropriately, so this doesn't break anything.
